src/Qir/Runtime/lib/QIR/QubitManager.cpp

```diff
@@ -0,0 +1,508 @@
+{
+    // We don't support true borrowing/returning at the moment.
+    return Allocate();
+}
+
+void CQubitManager::Borrow(Qubit* qubitsToBorrow, int32_t qubitCountToBorrow)
+{
+    // We don't support true borrowing/returning at the moment.
+    return Allocate(qubitsToBorrow, qubitCountToBorrow);
+}
+
+void CQubitManager::Return(Qubit qubit)
+{
+    // We don't support true borrowing/returning at the moment.
+    Release(qubit);
+}
+
+void CQubitManager::Return(Qubit* qubitsToReturn, int32_t qubitCountToReturn)
+{
+    // We don't support true borrowing/returning at the moment.
+    Release(qubitsToReturn, qubitCountToReturn);
+}
+
+void CQubitManager::Disable(Qubit qubit)
+{
+    FailIf(!IsValidQubit(qubit), "Qubit is not valid.");
+    QubitIdType id = QubitToId(qubit);
+
+    // We can only disable explicitly allocated qubits that were not borrowed.
+    FailIf(!IsExplicitlyAllocated(id), "Cannot disable qubit that is not explicitly allocated.");
+    sharedQubitStatusArray[id] = DisabledMarker;
+
+    disabledQubitCount++;
+    FailIf(disabledQubitCount <= 0, "Incorrect disabled qubit count.");
+    allocatedQubitCount--;
+    FailIf(allocatedQubitCount < 0, "Incorrect allocated qubit count.");
+}
+
+void CQubitManager::Disable(Qubit* qubitsToDisable, int32_t qubitCountToDisable)
+{
+    if (qubitCountToDisable == 0)
+    {
+        return;
+    }
+    FailIf(qubitCountToDisable < 0, "Cannot disable negative number of qubits.");
+    FailIf(qubitsToDisable == nullptr, "No array provided with qubits to be disabled.");
+
+    for (int32_t i = 0; i < qubitCountToDisable; i++)
+    {
+        Disable(qubitsToDisable[i]);
+    }
+}
+
+bool CQubitManager::IsValidQubit(Qubit qubit) const
+{
+    QubitIdType id = QubitToId(qubit);
+    if (id >= qubitCapacity)
+    {
+        return false;
+    }
+    if (id < 0)
+    {
+        return false;
+    }
+    return true;
+}
+
+bool CQubitManager::IsDisabledQubit(Qubit qubit) const
+{
+    return IsValidQubit(qubit) && IsDisabled(QubitToId(qubit));
+}
+
+bool CQubitManager::IsFreeQubit(Qubit qubit) const
+{
+    return IsValidQubit(qubit) && IsFree(QubitToId(qubit));
+}
+
+CQubitManager::QubitIdType CQubitManager::GetQubitId(Qubit qubit) const
+{
+    FailIf(!IsValidQubit(qubit), "Not a valid qubit.");
+    return QubitToId(qubit);
+}
+
+
+Qubit CQubitManager::CreateQubitObject(QubitIdType id)
+{
+    FailIf(id < 0 || id > INTPTR_MAX, "Qubit id is out of range.");
+    intptr_t pointerSizedId = static_cast<intptr_t>(id);
+    return reinterpret_cast<Qubit>(pointerSizedId);
+}
+
+void CQubitManager::DeleteQubitObject(Qubit qubit)
+{
+    // Do nothing. By default we store qubit Id in place of a pointer to a qubit.
+}
+
+CQubitManager::QubitIdType CQubitManager::QubitToId(Qubit qubit) const
+{
+    intptr_t pointerSizedId = reinterpret_cast<intptr_t>(qubit);
+    FailIf(pointerSizedId < 0 || pointerSizedId > std::numeric_limits<QubitIdType>::max(), "Qubit id is out of range.");
```

Comment:
Did you mean
```suggestion
    FailIf(pointerSizedId < 0 || pointerSizedId >= MaximumQubitCapacity, "Qubit id is out of range.");
```
or
```suggestion
    FailIf(pointerSizedId < 0 || pointerSizedId >= qubitCapacity, "Qubit id is out of range.");
```
?

Reply:
Again, the intention of this check is to verify that the value we are casting in the next line will fit into the variable of destination type.